File: src/main/java/com/jacuum/map/CellularMaps.java
package com.jacuum.map;

import com.jacuum.algo.Direction;
import java.util.*;

public final class CellularMaps implements Maps {

    private final int smoothingPasses;
    private final double fillRatio;

    public CellularMaps() {
        this.smoothingPasses = 5;
        this.fillRatio = 0.45;
    }

    @Override
    public GameMap generate(String hash, SizePreset size) throws Exception {
        long seed = seedFrom(hash);
        Random rng = new Random(seed);
        int w = size.width(), h = size.height();

        boolean[][] floor = initialFloor(rng, w, h);
        for (int i = 0; i < this.smoothingPasses; i++)
            floor = smooth(floor, w, h);

        floor = keepLargestRegion(floor, w, h);

        int[] start = centroidOfFloor(floor, w, h);
        return new GeneratedMap(hash, size, floor, start[0], start[1]);
    }

    private long seedFrom(String hash) {
        long h = 0xcbf29ce484222325L;
        for (char c : hash.toCharArray())
            h = (h ^ c) * 0x100000001b3L;
        return h;
    }

    private boolean[][] initialFloor(Random rng, int w, int h) {
        boolean[][] f = new boolean[h][w];
        for (int y = 1; y < h - 1; y++)
            for (int x = 1; x < w - 1; x++)
                f[y][x] = rng.nextDouble() > this.fillRatio;
        return f;
    }

    private boolean[][] smooth(boolean[][] f, int w, int h) {
        boolean[][] next = new boolean[h][w];
        for (int y = 1; y < h - 1; y++) {
            for (int x = 1; x < w - 1; x++) {
                int walls = 0;
                for (int dy = -1; dy <= 1; dy++)
                    for (int dx = -1; dx <= 1; dx++)
                        if (!f[y + dy][x + dx]) walls++;
                next[y][x] = walls < 5;
            }
        }
        return next;
    }

    private boolean[][] keepLargestRegion(boolean[][] floor, int w, int h) {
        boolean[][] visited = new boolean[h][w];
        List<List<int[]>> regions = new ArrayList<>();
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                if (floor[y][x] && !visited[y][x]) {
                    List<int[]> region = new ArrayList<>();
                    Queue<int[]> q = new ArrayDeque<>();
                    q.add(new int[]{x, y});
                    visited[y][x] = true;
                    while (!q.isEmpty()) {
                        int[] cur = q.poll();
                        region.add(cur);
                        for (Direction d : Direction.values()) {
                            int nx = cur[0] + d.dx(), ny = cur[1] + d.dy();
                            if (nx >= 0 && ny >= 0 && nx < w && ny < h
                                && floor[ny][nx] && !visited[ny][nx]) {
                                visited[ny][nx] = true;
                                q.add(new int[]{nx, ny});
                            }
                        }
                    }
                    regions.add(region);
                }
            }
        }
        if (regions.isEmpty()) {
            // Fallback: open 3x3 centre
            boolean[][] fallback = new boolean[h][w];
            int cx = w / 2, cy = h / 2;
            for (int dy = -1; dy <= 1; dy++)
                for (int dx = -1; dx <= 1; dx++)
                    fallback[cy + dy][cx + dx] = true;
            return fallback;
        }
        List<int[]> largest = regions.stream()
            .max(Comparator.comparingInt(List::size)).orElseThrow();
        boolean[][] result = new boolean[h][w];
        for (int[] cell : largest)
            result[cell[1]][cell[0]] = true;
        return result;
    }

    private int[] centroidOfFloor(boolean[][] floor, int w, int h) {
        long sx = 0, sy = 0, count = 0;
        for (int y = 0; y < h; y++)
            for (int x = 0; x < w; x++)
                if (floor[y][x]) { sx += x; sy += y; count++; }
        if (count == 0) return new int[]{w / 2, h / 2};
        int cx = (int) (sx / count), cy = (int) (sy / count);
        // Snap to nearest floor tile
        int best = Integer.MAX_VALUE;
        int[] result = new int[]{cx, cy};
        for (int y = 0; y < h; y++) {
            for (int x = 0; x < w; x++) {
                if (floor[y][x]) {
                    int dist = (x - cx) * (x - cx) + (y - cy) * (y - cy);
                    if (dist < best) { best = dist; result = new int[]{x, y}; }
                }
            }
        }
        return result;
    }
}

File: src/main/kotlin/com/jacuum/map/CellularMaps.kt
package com.jacuum.map

import com.jacuum.algo.Direction
import java.util.ArrayDeque
import java.util.Random

class CellularMaps : Maps {

    private val smoothingPasses = 5
    private val fillRatio = 0.45

    override fun generate(hash: String, size: SizePreset): GameMap {
        val rng = Random(seedFrom(hash))
        val w = size.width; val h = size.height
        var floor = initialFloor(rng, w, h)
        repeat(smoothingPasses) { floor = smooth(floor, w, h) }
        floor = keepLargestRegion(floor, w, h)
        val (sx, sy) = centroidOfFloor(floor, w, h)
        return GeneratedMap(hash, size, floor, sx, sy)
    }

    private fun seedFrom(hash: String): Long {
        var h = -3750763034362895579L
        for (c in hash) h = (h xor c.code.toLong()) * 1099511628211L
        return h
    }

    private fun initialFloor(rng: Random, w: Int, h: Int): Array<BooleanArray> {
        val f = Array(h) { BooleanArray(w) }
        for (y in 1 until h - 1)
            for (x in 1 until w - 1)
                f[y][x] = rng.nextDouble() > fillRatio
        return f
    }

    private fun smooth(f: Array<BooleanArray>, w: Int, h: Int): Array<BooleanArray> {
        val next = Array(h) { BooleanArray(w) }
        for (y in 1 until h - 1)
            for (x in 1 until w - 1) {
                var walls = 0
                for (dy in -1..1) for (dx in -1..1) if (!f[y + dy][x + dx]) walls++
                next[y][x] = walls < 5
            }
        return next
    }

    private fun keepLargestRegion(floor: Array<BooleanArray>, w: Int, h: Int): Array<BooleanArray> {
        val visited = Array(h) { BooleanArray(w) }
        val regions = mutableListOf<List<Pair<Int, Int>>>()
        for (y in 0 until h) for (x in 0 until w) {
            if (floor[y][x] && !visited[y][x]) {
                val region = mutableListOf<Pair<Int, Int>>()
                val q = ArrayDeque<Pair<Int, Int>>()
                q.add(x to y); visited[y][x] = true
                while (q.isNotEmpty()) {
                    val (cx, cy) = q.poll(); region.add(cx to cy)
                    for (d in Direction.values()) {
                        val nx = cx + d.dx(); val ny = cy + d.dy()
                        if (nx in 0 until w && ny in 0 until h && floor[ny][nx] && !visited[ny][nx]) {
                            visited[ny][nx] = true; q.add(nx to ny)
                        }
                    }
                }
                regions.add(region)
            }
        }
        if (regions.isEmpty()) {
            val fallback = Array(h) { BooleanArray(w) }
            val cx = w / 2; val cy = h / 2
            for (dy in -1..1) for (dx in -1..1) fallback[cy + dy][cx + dx] = true
            return fallback
        }
        val largest = regions.maxByOrNull { it.size }!!
        val result = Array(h) { BooleanArray(w) }
        for ((rx, ry) in largest) result[ry][rx] = true
        return result
    }

    private fun centroidOfFloor(floor: Array<BooleanArray>, w: Int, h: Int): Pair<Int, Int> {
        var sx = 0L; var sy = 0L; var count = 0L
        for (y in 0 until h) for (x in 0 until w) if (floor[y][x]) { sx += x; sy += y; count++ }
        if (count == 0L) return w / 2 to h / 2
        val cx = (sx / count).toInt(); val cy = (sy / count).toInt()
        var best = Int.MAX_VALUE; var result = cx to cy
        for (y in 0 until h) for (x in 0 until w) if (floor[y][x]) {
            val dist = (x - cx) * (x - cx) + (y - cy) * (y - cy)
            if (dist < best) { best = dist; result = x to y }
        }
        return result
    }
}
